### PR TITLE
Optimize DPO recipe - precomputing reference model log probabilites

### PR DESCRIPTION
**Pull request by yash12khandelwal**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
The primary purpose of this PR is to add support for precomputing reference log probabilities when using DPO. This would make the overall training faster by removing the redundant computation across epochs. 
- `CustomPreferenceDataset` - This file is a modification of the Preference Dataset that allows the storage of the reference log probabilities along with the data. Every get-item call would return a dictionary of input_ids, labels and the reference model chosen and rejected log probabilities.
- `padded_collate_dpo` - Modified this function to return the precomputed log probabilities too along with the inputs and labels.
- `lora_dpo_distributed` - Added the support to precompute the reference log probabilities during data setup. For computing losses, the batch item can return the precomputed values saving compute. Implementation is inspired from the [Hugging Face trl repository DPO implementation](https://github.com/huggingface/trl/blob/763738f457f283270772ac9bd5b3e4027fd424d5/trl/trainer/dpo_trainer.py#L708).

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/configs/llama3_1/8B_lora_dpo_precompute.yaml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  ]
+  recipe_checkpoint: null
+  output_dir: ${output_dir}
+  model_type: LLAMA3
+resume_from_checkpoint: False
+save_adapter_weights_only: False
+
+# Dataset and Sampler
+dataset:
+  _component_: torchtune.datasets.stack_exchange_paired_dataset
```
Comment (aashay-sarvam) [inline recipes/configs/llama3_1/8B_lora_dpo_precompute.yaml:54]: Component needs change
